prusti-encoder/src/encoders/ty/pure.rs

```diff
@@ -0,0 +1,591 @@
+    }
+}
+
+impl<'vir> Deref for AdtBuilder<'vir> {
+    type Target = TyPureBuilder<'vir>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+pub(crate) struct TyPureBuilder<'vir> {
+    pub(crate) vcx: &'vir vir::VirCtxt<'vir>,
+    name: &'vir str,
+    generics: Option<Vec<vir::LocalDeclTyVal<'vir>>>,
+    domain_ident: vir::DomainIdnSnap<'vir>,
+    self_type: vir::TypeSnap<'vir>,
+    unreachable_to_snap: FunctionIdn<'vir, vir::ManyTyVal, vir::Snap>,
+    pub(super) params: GenericParams<'vir>,
+    data: BuilderData<'vir>,
+}
+
+pub enum BuilderData<'vir> {
+    Adt(AdtBuilderData<'vir>),
+    Domain(DomainBuilderData<'vir>),
+    None,
+}
+
+#[derive(Default)]
+pub(crate) struct AdtBuilderData<'vir> {
+    constructors: Vec<vir::AdtConstructor<'vir>>,
+    discr_fn: Option<DiscrFnBuilder<'vir>>,
+}
+
+#[derive(Default)]
+pub(crate) struct DomainBuilderData<'vir> {
+    axioms: Vec<vir::DomainAxiom<'vir>>,
+    functions: Vec<vir::DomainFunction<'vir>>,
+}
+
+#[derive(Clone, Copy)]
+enum DiscrFnBuilder<'vir> {
+    Building {
+        param: vir::LocalDeclCSnap<'vir>,
+        recv: vir::ExprCSnap<'vir>,
+        acc: vir::ExprCSnap<'vir>,
+    },
+    Built(vir::Function<'vir>),
+}
+
+impl<'vir> TyPureBuilder<'vir> {
+    pub(crate) fn new<E: TaskEncoder>(deps: &mut TaskEncoderDependencies<'vir, E>, vcx: &'vir vir::VirCtxt<'vir>, ty: RustTy<'vir>) -> Self {
+        let params = deps.require_dep::<GenericParamsEnc>(ty.params).unwrap();
+        let name = vir::vir_format!(vcx, "s_{}", ty.name());
+        let domain_ident = DomainIdnSnap::new(vir::ViperIdent::new(name));
+        let self_type = domain_ident();
+        let unreachable_to_snap = FunctionIdn::new(
+            vir::ViperIdent::new(vir::vir_format!(vcx, "{name}_unreachable")),
+            params.ty_args(),
+            self_type,
+        );
+        TyPureBuilder {
+            vcx,
+            name,
+            generics: None,
+            domain_ident,
+            self_type,
+            unreachable_to_snap,
+            params,
+            data: BuilderData::None,
+        }
+    }
+
+    pub(crate) fn set_generics(&mut self, generics: Vec<vir::LocalDeclTyVal<'vir>>) {
+        self.generics = Some(generics);
+    }
+
+    pub(crate) fn self_type(&self) -> vir::TypeCSnap<'vir> {
+        self.self_type.downcast_ty()
+    }
+
+    pub(crate) fn output_ref(&self) -> TyPureRef<'vir> {
+        TyPureRef {
+            domain: self.domain_ident.cast_ty(),
+            unreachable_to_snap: self.unreachable_to_snap,
+        }
+    }
+
+    pub(crate) fn set_domain_builder(&mut self) -> &mut DomainBuilder<'vir> {
+        match &mut self.data {
+            BuilderData::Adt(_) => panic!("already an ADT builder"),
+            BuilderData::Domain(_) => {}
+            data@BuilderData::None => {
+                *data = BuilderData::Domain(DomainBuilderData::default());
+            }
+        }
+        // SAFETY: `DomainBuilder` is repr transparent
```

Comment:
It's not, though. In general I don't see the need for these tricks (also for the slice of lazy Rust types), since they seem like (potentially erroneous) micro-optimisations.

Reply:
Ah I marked the Adt one, but forgot about this one. This specifically is not an optimisation in any way - there just simply isn't any other way to go from `&mut TyPureBuilder` to `&mut DomainBuilder`. It's just a wrapper which guarantees that the inner `data` is `BuilderData::Domain` and so we can expose the various functions for building the domain. I don't see a nicer way to do it (even slower ones).